src/Control/Carrier/Simple.hs

```diff
@@ -0,0 +1,115 @@
+
+Should your performance requirements change, it is generally possible to reformulate the interpretation of an effect with 'Algebra' without breaking client code.
+
+This module is based on (and almost entirely API-compatible with) work done at [FOSSA](https://fossa.com) and released as part of their [fossa-cli](https://github.com/fossas/fossa-cli) project. We are grateful for their contributions.
+
+@since 1.2.0.0
+
+= Defining effects with 'Simple'
+
+Effect types compatible with 'Simple' take fewer type parameters than 'Algebra'-compatible types. Here is the classic teletype example. Note that we define a type synonym for @Simple TeletypeF@ so that the implementation details of this effect do not leak into clients.
+
+@
+  data TeletypeF a where
+    Read :: TeletypeF String
+    Write :: String -> TeletypeF ()
+
+  type Teletype = Simple TeletypeF
+@
+
+Whereas an 'Algebra'-compatible @TeletypeF@ effect would take an @m@ monadic parameter and @k@ continuation parameter, those are elided in the simple formulation, as this formulation is intended to hide the details of monads and their associated state parameters.
+
+Effects are constructed by wrapping these effect types with 'sendSimple':
+
+@
+  read :: Has Teletype sig m => m String
+  read = sendSimple Read
+
+  write :: Has Teletype sig m => String -> m ()
+  write = sendSimple . Write
+@
+
+Defining an interpretation function is easy:
+
+@
+  runTeletypeIO :: MonadIO m => TeletypeF a -> m a
+  runTeletypeIO t = liftIO $ case t of
+    Read -> getLine
+    Write s -> putStrLn s
+@
+
+-}
+
+module Control.Carrier.Simple
+( -- * Simple carrier
+  interpret
+, SimpleC(..)
+, HandlerFor(..)
+  -- * Stateful interpretation
+, interpretState
+, SimpleStateC
+, module Control.Effect.Simple
+) where
+
+import Control.Algebra
+import Control.Applicative
+import Control.Carrier.Reader
+import Control.Carrier.State.Strict
+import Control.Monad (MonadPlus)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Control.Monad.Fix (MonadFix)
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Trans.Class
+import Control.Effect.Simple
+
+-- | Run a 'SimpleC' carrier wrapping an effect type @eff@ with a function that interprets values of type @eff x@ into values associated with some underlying monad @m@.
+interpret :: (forall x. eff x -> m x) -> SimpleC eff m a -> m a
+interpret f = runReader (HandlerFor f) . runSimpleC
+
+newtype SimpleC eff m a = SimpleC { runSimpleC :: ReaderC (HandlerFor eff m) m a }
+  deriving (Applicative, Alternative, Functor, Monad, Fail.MonadFail, MonadFix, MonadIO, MonadPlus, MonadUnliftIO)
+
+instance MonadTrans (SimpleC eff) where
+  lift = SimpleC . lift
+
+-- | An existential wrapper for an effect handler function (a natural transformation between @eff@ and @m@).
+data HandlerFor eff m = HandlerFor (forall a. eff a -> m a)
```

Comment:
1. This isn't an existential wrapper, `forall` inside the constructor is a universal quantifier.
2. `newtype`.